package/samplers/ctpe/README.md

```diff
@@ -0,0 +1,89 @@
+---
+author: Shuhei Watanabe
+title: c-TPE; Tree-structured Parzen Estimator with Inequality Constraints for Expensive Hyperparameter Optimization
+description: The optimizer that reproduces the algorithm described in the paper ``c-TPE; Tree-structured Parzen Estimator with Inequality Constraints for Expensive Hyperparameter Optimization''.
+tags: [sampler, tpe, c-tpe, paper, research]
+optuna_versions: [v4.1.0]
+license: MIT License
+---
+
+## Abstract
+
+This package aims to reproduce the TPE algorithm used in the paper published at IJCAI'23:
+
+- [c-TPE: Tree-structured Parzen Estimator with Inequality Constraints for Expensive Hyperparameter Optimization](https://arxiv.org/abs/2211.14411)
+
+The default parameter set of this sampler is the recommended setup from the paper and the experiments in the paper can also be reproduced by this sampler.
+
+## Class or Function Names
+
+- cTPESampler
+
+Although most arguments in [`TPESampler`](https://optuna.readthedocs.io/en/stable/reference/samplers/generated/optuna.samplers.TPESampler.html) are also available for `cTPESampler`, the original paper uses the default arguments of this sampler.
+However, `constraints_func` must be provided to use this sampler.
+For the reproducibility purpose, please set a fixed `seed`.
+
+## Installation
+
+The version constraint of this package is Optuna v4.0.0 or later.
+
+```shell
+# The requirement is only Optuna.
+$ pip install optuna
+
+# You can also optionally install as follows:
+$ pip install -r https://hub.optuna.org/samplers/ctpe/requirements.txt
+```
+
+## Example
+
+This sampler is the official implementation of [the c-TPE paper](https://arxiv.org/abs/2211.14411).
+
+```python
+from __future__ import annotations
+
+import numpy as np
+import optuna
+import optunahub
+
+
+def objective(trial: optuna.Trial) -> float:
+    x = trial.suggest_float("x", 0.0, 2 * np.pi)
+    y = trial.suggest_float("y", 0.0, 2 * np.pi)
+    return float(np.sin(x) + y)
+
+
+def constraints(trial: optuna.trial.FrozenTrial) -> tuple[float]:
+    x = trial.params["x"]
+    y = trial.params["y"]
+    c = float(np.sin(x) * np.sin(y) + 0.95)
+    trial.set_user_attr("c", c)
+    return (c, )
+
+
+sampler = optunahub.load_module(package="samplers/ctpe").cTPESampler(constraints_func=constraints)
+study = optuna.create_study(sampler=sampler)
+study.optimize(objective, n_trials=30)
+print(study.best_trials)
+
+```
+
```

Comment:
It is helpful to describe how to define multiple constraints in cTPE.

Reply:
I did not exactly get what you meant, but isn't it possible to simply return two constraint values in `constraints` as this sampler takes the same interface as Optuna, or did you mean I should defer to any Optuna API docs?

Reply:
Providing an example with multiple constraints like the one below is helpful.
However, adding a link to [the docs](https://optuna.readthedocs.io/en/stable/faq.html#how-can-i-optimize-a-model-with-some-constraints) might be enough.

```python
def constraints(trial: optuna.trial.FrozenTrial) -> tuple[float]:
    x = trial.params["x"]
    y = trial.params["y"]
    c1 = float(np.sin(x) * np.sin(y) + 0.95)
    trial.set_user_attr("c1", c1)
    c2 = ...
    trial.set_user_attr("c2", c2)
    return c1, c2
```

Reply:
Gotcha, I added the reference!